lib/Language/Slicer/Absyn.hs

```diff
@@ -105,8 +108,28 @@ data Type = IntTy | BoolTy | UnitTy | StringTy
           | TyVar TyVar
           -- References
           | RefTy Type
+          -- Exception type
+          | ExnTy
           -- Trace types
           | TraceTy Type
-            deriving (Show, Eq, Ord, Generic, NFData)
+            deriving (Show, Generic, NFData)
+
+-- We need a hand-written instance of Eq, because ExnTy is equal to every type
+instance Eq Type where
+    ExnTy        == _              = True
```

Comment:
This is a little strange.  I suspect the right thing to do in the presence of exceptions is to actually implement some (maybe bidirectional) type inference.  Then "raise e" will have type alpha where alpha is a  fresh type variable.  We already have type variables for recursive types.